In [1]:
%%time

#source: https://towardsdatascience.com/how-fast-is-c-compared-to-python-978f18f474c7

def convert(c):
    if (c == 'A'): return 'C'
    if (c == 'C'): return 'G'
    if (c == 'G'): return 'T'
    if (c == 'T'): return 'A'
    
print("Start")

opt = "ACGT"
s = ""
s_last = ""
len_str = 13

for i in range(len_str):
    s += opt[0]

for i in range(len_str):
    s_last += opt[-1]

pos = 0
counter = 1
while (s != s_last):
    counter += 1
    # You can uncomment the next line to see all k-mers.
    # print(s)
    change_next = True
    for i in range(len_str):
        if (change_next):
            if (s[i] == opt[-1]):
                s = s[:i] + convert(s[i]) + s[i+1:]
                change_next = True
            else:
                s = s[:i] + convert(s[i]) + s[i+1:]
                break

# You can uncomment the next line to see all k-mers.
# print(s)
print("Number of generated k-mers: {}".format(counter))
print("Finish!")

Start
Number of generated k-mers: 67108864
Finish!
CPU times: user 1min 3s, sys: 7.62 ms, total: 1min 3s
Wall time: 1min 3s


In [2]:
# pip3 install cython wurlitzer
%load_ext Cython
%load_ext wurlitzer

In [5]:
%%cython

'''
slightly improved code,
using dictionary instead of a function
removed two unnecessary loops , an unnecessary condition
declared variable types beforehanded and compiled it like C/C++
'''

from libc.stdio cimport printf

def do():
    cdef int counter = 1
    cdef int len_str = 13
    cdef str opt = "ACGT"
    cdef str s = opt[0]*len_str
    cdef str s_last = opt[-1]*len_str
    cdef int i
    cdef dict convert = {
                    'A':'C',
                    'C':'G',
                    'G':'T',
                    'T':'A'}
    
    printf("Start")
    
    while (s != s_last):
        counter += 1
        for i in range(len_str):
            if (s[i] == opt[-1]): 
                s = s[:i] + convert[s[i]] + s[i+1:]
            else:
                s = s[:i] + convert[s[i]] + s[i+1:]
                break
                
    printf("\nNumber of generated k-mers:%d",counter)
    printf("\nFinish!")

In [6]:
%%time
do()

CPU times: user 10.1 s, sys: 24.1 ms, total: 10.1 s
Wall time: 10.1 s
Start
Number of generated k-mers:67108864
Finish!

***Pure Python code took 1 min 3 seconds i.e 63 seconds, while Cythonized code took 10.1 seconds only that is 6X speed up,***

***so Python can be as fast as C/C++***